### PR TITLE
MLM Pre-masked

### PR DESCRIPTION
**Pull request by zphang**
Adding two generic tasks, both supporting cases where we're doing MLM over pre-specified masked inputs.

* `mlm_premasked`: Input is text, masks specified as char spans. The tokenization is not 100% correct (we're basically tokenizing each segment separately, which is not the same as tokenizing the whole sequence), but assuming this will be used in fine-tuning, this should be fine.
* `mlm_pretokenized`: Input is list of tokens (has to be specific to tokenizer) masks are specific as token spans. This provides an option for when tokenization needs to be 100% correct.

### CODE REVIEW

File: jiant/tasks/evaluate/core.py [hunk 1]
```diff
@@ -141,6 +145,21 @@ def compute_metrics_from_accumulator(
         return self.compute_metrics_from_preds_and_labels(preds=preds, labels=labels)
 
 
+class MLMPremaskedAccumulator(BaseAccumulator):
+    def __init__(self):
+        self.logits_list = []
+
+    def update(self, batch_logits, batch_loss, batch, batch_metadata):
+        batch_size = len(batch)
+        non_masked = batch.masked_lm_labels.cpu().numpy() != mlm_template.NON_MASKED_TOKEN_LABEL_ID
```
Comment (jeswan) [inline jiant/tasks/evaluate/core.py:154]: clarify what non_masked is here

File: jiant/tasks/lib/templates/mlm_premasked.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        )
+
+
+@dataclass
+class TokenizedExample(BaseTokenizedExample):
+    guid: str
+    masked_tokens: List
+    label_tokens: List
+
+    def featurize(self, tokenizer, feat_spec):
```
Comment (jeswan) [inline jiant/tasks/lib/templates/mlm_premasked.py:61]: Discussed offline - potential for a test here to ensure featurization is correct

File: jiant/tasks/lib/templates/mlm_premasked.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+
+@dataclass
+class Example(BaseExample):
+    guid: str
+    text: str
+    # Spans over char indices
+    masked_spans: List[ExclusiveSpan]
+
+    def tokenize(self, tokenizer):
```
Comment (jeswan) [inline jiant/tasks/lib/templates/mlm_premasked.py:28]: Discussed offline - potential for a test here to ensure tokenization is correct

File: jiant/tasks/lib/mlm_pretokenized.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            label_tokens += tokenized_masked_text
+            curr = end
+        if curr < len(self.tokenized_text):
+            tokenized_text = self.tokenized_text[curr:]
+            masked_tokens += tokenized_text
+            label_tokens += [tokenizer.pad_token] * len(tokenized_text)
+
+        return TokenizedExample(
+            guid=self.guid, masked_tokens=self.tokenized_text, label_tokens=label_tokens,
+        )
```
Comment (jeswan) [inline jiant/tasks/lib/mlm_pretokenized.py:45]: is this the same as tokenize() below? if so, could be refactored

Comment (zphang) [reply jiant/tasks/lib/mlm_pretokenized.py:45]: This is the portion where we handle the tokenization wrt masking, so it's different from other tasks.

File: jiant/tasks/lib/templates/mlm_premasked.py (hunk 2 above)
Comment (zphang) [reply jiant/tasks/lib/templates/mlm_premasked.py:61]: Added

**Review by jeswan — APPROVED** [review]
